fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LucenOnlineIndexingTest.java

```diff
@@ -0,0 +1,604 @@
+            }
+            metaDataBuilder.removeIndex(TextIndexTestUtils.SIMPLE_DEFAULT_NAME);
+            TextIndexTestUtils.addRecordTypePrefix(metaDataBuilder);
+        };
+        try (final FDBRecordContext context = openContext()) {
+            recordStore = LuceneIndexTestUtils.openRecordStore(context, path, hook);
+            for (Index index: indexes) {
+                recordStore.markIndexDisabled(index).join();
+            }
+            context.commit();
+        }
+        try (final FDBRecordContext context = openContext()) {
+            recordStore = LuceneIndexTestUtils.openRecordStore(context, path, hook);
+            for (int i = 0; i < numRecords; i ++) {
+                long docId = docIds[i];
+                recordStore.saveRecord(createSimpleDocument(docId, randomText(rn), randomGroup(rn)));
+            }
+            context.commit();
+        }
+        // overwrite some records (to enforce merge), write others as new
+        try (final FDBRecordContext context = openContext()) {
+            recordStore = LuceneIndexTestUtils.openRecordStore(context, path, hook);
+            for (int i = 0; i < middle; i ++) {
+                long docId = docIds[i];
+                recordStore.saveRecord(createSimpleDocument(docId, randomText(rn), randomGroup(rn)));
+            }
+            for (int i = middle; i < numRecords; i ++) {
+                long docId = docIds[numRecords + i];
+                recordStore.saveRecord(createSimpleDocument(docId, randomText(rn), randomGroup(rn)));
+            }
+            context.commit();
+        }
+        // build the index ..
+        try (final FDBRecordContext context = openContext()) {
+            recordStore = LuceneIndexTestUtils.openRecordStore(context, path, hook);
+            try (OnlineIndexer indexBuilder = OnlineIndexer.newBuilder()
+                    .setRecordStore(recordStore)
+                    .setTargetIndexes(indexes)
+                    .setLimit(transactionLimit)
+                    .setIndexingPolicy(OnlineIndexer.IndexingPolicy.newBuilder()
+                            .setDeferLuceneMergeDuringIndexing(preventMergeDuringIndexing)
+                            .build())
+                    .build()) {
+                for (Index index: indexes) {
+                    assertTrue(recordStore.isIndexDisabled(index));
+                }
+                indexBuilder.buildIndex(true);
+            }
+        }
+        // .. and assert readable mode
+        try (final FDBRecordContext context = openContext()) {
+            recordStore = LuceneIndexTestUtils.openRecordStore(context, path, hook);
+            for (Index index: indexes) {
+                assertTrue(recordStore.isIndexReadable(index));
+            }
+        }
+        // assert number of segments
+        for (Index index: indexes) {
+            String[] allFiles = listFiles(index);
+            assertTrue(allFiles.length < 12);
```

Comment:
I may have missed it, but would it make sense to test the logic by creating multiple stores - one with the defer option set to `TRUE` and one with the defer option as `FALSE`, add a few documents to the first, check the number of segments, add more documents to the other and assert that the number of segments is smaller?

Reply:
That can be done.

Reply:
✅